EIPS/eip-rich-site-proposed-contract-metadata.md

```diff
@@ -0,0 +1,100 @@
+
+Additionally, this method serves to help populate the wallet with trustworthy information about what it holds. It reduces the need to rely on any central services for indexing (enhancing performance and privacy), and eliminates dangers that come from index-based automatic asset detection, like receiving harrassment, or [Airdrop scams](https://medium.com/metamask/phisher-watch-airdrop-scams-82eea95d9b2a).
+
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Wallets may integrate the ability to propose names for contracts, methods, and parameters into any method that touches an address, so the wallet can be expanded with this metadata at runtime. These methods include `eth_sendTransaction`, and `eth_signTypedData` (and its variants).
+
+Benefits
+  - No need for any central server
+  - No registry with collisions
+  - Works on all chains
+  - Includes parameter names as well as method and contract names.
+  - Could be easily added to any existing dapp
+  - If every dapp a user touches uses this, then no phishing site could get the user to give up one of those assets without seeing a legible confirmation.
+  - Can even be implemented on a cold/offline or hardware wallet.
+
+### Implementation
+Any method that involves an address MAY include a new OPTIONAL parameter `proposedContracts`, which will include an additional prompt to the user as part of the confirmation flow, asking if they'd like to trust the site for this information.
+
+`contractId` would be some kind of cross chain address format, likely [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md).
+`ContractData` would be some kind of type that is enough to render transaction names for methods called.
+  - Today TX Insight uses a full Truffle Decoder output to render this, which includes lots of compiler info that we would not need here.
+```typescript
+type TxOptionsEnhancement = {
+  proposedContracts: {
+    [contractId: CrossChainAddress]: {
+      proposedName: string,
+      abi?: Abi,
+      contractData?: ContractData
+    }
+  }
+}
+```
+Wallets MUST store the recommending site & time along with this data for forensic purposes.
+The user SHOULD be able to adjust the proposed names in the confirmation, to avoid collisions, and help personalize their local context.
+Any subsequent confirmations that are displaying the proposed contracts' names SHOULD show the proposed names.
+  - A mouse hover may provide the metadata about when this name was proposed.
+Any subsequent transaction confirmations that are displaying transactions to the proposed contracts SHOULD use the proposed data to render those transactions as readably as possible.
+Additional optional parameters MAY be added to the `proposedContracts` options object to allow new kinds of enhancement in the future.
```

Comment:
This feels like it should be in the Specification section, rather than the Reference Implementation section.  Note that the Reference Implementation section is optional, so feel free to leave it out if you don't have any content appropriate for it.

Reply:
Moved in latest commit. I hadn't noticed the Reference Implementation header.